### PR TITLE
73 water vapor notebook

### LINKED ISSUE #73: Add a example notebook for water vapor
similar to SST, we want an example notebook for hourly water vapor prediction. The goal is to check if the current model and modules works for water vapor.

### PR DESCRIPTION
**Pull request by rogerkuou**
Fix #73 

Added check for empty dataset
Init loss in predict as a tensor
In the example notebook down sample landsurface mask by the factor of 2

### CODE REVIEW

**Comment by rogerkuou** [comment]
Hi @SarahAlidoost , while working on #73, I got an error in [the notebook](https://github.com/ESMValGroup/ClimaNet/blob/5b3d0df970b6ede1613b37a30e164af17ab6d8d1/notebooks/example_hourly_watervapor.ipynb). This notebooks is adapted from the SST hourly one.

It fails on `average_loss.item()`, which hints that `average_loss` is a dictionany. Do you have clue on this? I am running with the CPU mode instead of GPU.

**Comment by rogerkuou** [comment]
Hi @SarahAlidoost , can you review this notwbook when you have time?

The error was caused by an empty dataset. I added a check to predict, and initiated loss as an zero Tensor.

Another thing maybe we need to pay attention, is the water wapor data is downsampled by the factor of 2 comparing to SST. This requires some attention to lsm data. In the notebook I used an average window to downsample the lsm, and set <0.5 to 0, and >=0.5 to 1. We can discuss if this makes sense.

File: climanet/predict.py [hunk 1]
```diff
@@ -116,7 +120,7 @@ def predict_monthly_var(
 
     with torch.inference_mode():
         idx = 0
-        average_loss = 0.0
+        average_loss = torch.tensor(0.0, device=device)
```
Comment (SarahAlidoost) [inline climanet/predict.py]: As you explained the error was related to the dataset, so no need for this change. Please keep `average_loss = 0.0`.

**Review by SarahAlidoost — COMMENTED** [review]
@rogerkuou Thanks! 👍 I added one comment on the changed code. Levante is not available today and I dont have data locally. I will check the notebook later. About the spatial resolution of water vapor, I expected era5 data have the same spatial resolution; this is something to check in our next update meeting. I couldnot also check the era5 data as CDS is also on maintenance.

File: climanet/predict.py (hunk 1 above)
Comment (rogerkuou) [reply climanet/predict.py]: fixed!

**Comment by SarahAlidoost** [comment]
> Hi @SarahAlidoost , can you review this notwbook when you have time?
> 
> The error was caused by an empty dataset. I added a check to predict, and initiated loss as an zero Tensor.
> 
> Another thing maybe we need to pay attention, is the water wapor data is downsampled by the factor of 2 comparing to SST. This requires some attention to lsm data. In the notebook I used an average window to downsample the lsm, and set <0.5 to 0, and >=0.5 to 1. We can discuss if this makes sense.

Indeed there are a few things about water vapor data to check:
- the spatial resolution of data on levante is 0.5. The water vapor data is available [on cds](https://cds.climate.copernicus.eu/datasets/reanalysis-era5-single-levels?tab=download) with resolution 0.25 
- data is cut between (-80, 80) in latitude dimension
- the latitudes are not increasing they are from (80, -80) while longitude are i.e. (-180, 180)

**Comment by rogerkuou** [comment]
> > Hi @SarahAlidoost , can you review this notwbook when you have time?
> > The error was caused by an empty dataset. I added a check to predict, and initiated loss as an zero Tensor.
> > Another thing maybe we need to pay attention, is the water wapor data is downsampled by the factor of 2 comparing to SST. This requires some attention to lsm data. In the notebook I used an average window to downsample the lsm, and set <0.5 to 0, and >=0.5 to 1. We can discuss if this makes sense.
> 
> Indeed there are a few things about water vapor data to check:
> 
> * the spatial resolution of data on levante is 0.5. The water vapor data is available [on cds](https://cds.climate.copernicus.eu/datasets/reanalysis-era5-single-levels?tab=download) with resolution 0.25
> * data is cut between (-80, 80) in latitude dimension
> * the latitudes are not increasing they are from (80, -80) while longitude are i.e. (-180, 180)

During the meeting in July 24, we discussed the issues with Axel:

- The resolution is coming from the fact the satellite observations on water vapor has resolution with 0.5 degress. Hence it does not add value to go to -.25 degree resolution
- The data cut and lat coordinate order comes from how regredding is set. Axel is working on generating a new version of dataset with (-90, 90) and ascending ordering.

@SarahAlidoost shall we consider merging this PR for now? My motivation is that this PR also renames example notebooks to distinguish sst and water vapor. It can set good templates for other PRs.

**Comment by rogerkuou** [comment]
Updated on 20260814
Ihave confirmed with Axel that the satellite data's coverage is (-80, 80), and the resolution is 0.5 deg. Axel prefers not to create a new set of data, but use the current version. This means for the order of lat we can flip it by ourselves.

About the masking, I found CDS always provide 0.25 degrees. If using esemble mean to download, the downloaded mask are floating values. I check with xel he is fine with thresholding the floating mask with 0.5 values. Maybe let's do that and create another mask file matching the Water vapor data?

What do you think @SarahAlidoost and @meiertgrootes ?

**Comment by rogerkuou** [comment]
Hi @SarahAlidoost, in this PR I made example notebooks for watervapor. I executed the daily one locally and the hourly one on Levante.

I observed that the validation loss does not reduce, and the training stops with early stops. This is especially significant in the hourly training. I have tried using different month (I started with Jan, then switched to April. It the same as SST setting, three identical month from 2020-2022 for training, validation and testing), and reduced the learning rate to half (just in case if we are missing optimal). The validation loss curve stays flat. I hope for a full training with multiple month can help. But I will wait till your test with #92 

(Updated on SEP 10)
Reagrding the small trial of validate/test only on training data, see this [GH Gist](https://gist.github.com/rogerkuou/d3bc63fd444129d71a4532b4d907f0f2). It should the traning loss continue to drop, and the prediction results are significantly improved comparing to baseline.
The "artificial oscillation" in the daily notebook is caused by an uncleared training history, where at same epoch there are multiple loss. To avoid conflicts, I solved this in #95. By removing caches and re-run we do not have this anymore. See the [updated notebook](https://github.com/ESMValGroup/ClimaNet/blob/52971a6d7e6b21419e0eae00e85f9223a9b5d6ac/notebooks/watervapor/training_daily_watervapor.ipynb).

I merged your fix in #92 to this PR, to include your fix for the bug. So it will make sense to review this after merging #92.

**Comment by rogerkuou** [comment]
Hi @SarahAlidoost , I brought this branch up to date. See my [previous comment](https://github.com/ESMValGroup/ClimaNet/pull/74#issuecomment-5584533969) for details.

Meanwhile, following our discussion today, I will make another test on the training dataset to see if the training is able to overfit on the training data. This can validate if the problem is actually training on one month

File: notebooks/sst/training_hourly_sst.ipynb [hunk 2]
```diff
@@ -50,7 +50,11 @@
     "\n",
     "monthly_data = xr.open_mfdataset(data_folder / f\"202001_mon_ERA5dc_full_{var_name}.nc\")\n",
     "monthly_data_validation = xr.open_mfdataset(data_folder / f\"202101_mon_ERA5dc_full_{var_name}.nc\")\n",
+<<<<<<< HEAD:notebooks/sst/training_hourly_sst.ipynb
+    "monthly_data_test = xr.open_mfdataset(data_folder / f\"202001_mon_ERA5dc_full_{var_name}.nc\")\n",
+=======
     "monthly_data_test = xr.open_mfdataset(data_folder / f\"202201_mon_ERA5dc_full_{var_name}.nc\")\n",
+>>>>>>> main:notebooks/example_hourly.ipynb
```
Comment (SarahAlidoost) [inline notebooks/sst/training_hourly_sst.ipynb]: these are leftover from merging. Could you fix them?

File: notebooks/sst/training_hourly_sst.ipynb [hunk 3]
```diff
@@ -476,7 +480,11 @@
     }
    ],
    "source": [
+<<<<<<< HEAD:notebooks/sst/training_hourly_sst.ipynb
+    "spatial_patch_size = monthly_da.shape[1:]  # the whole dataset \n",
+=======
     "spatial_crop_size = monthly_da.shape[1:]  # the whole dataset \n",
+>>>>>>> main:notebooks/example_hourly.ipynb
```
Comment (SarahAlidoost) [inline notebooks/sst/training_hourly_sst.ipynb]: same here.

File: scripts/tuning.py [hunk 4]
```diff
@@ -85,7 +85,8 @@
         "dataloader_persistent_workers": True,
         "dataloader_multiprocessing_context": None,  # load_lazy is False
         "num_epoch": 100,
-        "max_concurrent_trials": args.num_nodes * 2,  # less than GPUs per node (4) avoid OOM
+        "max_concurrent_trials": args.num_nodes
+        * 2,  # less than GPUs per node (4) avoid OOM
```
Comment (SarahAlidoost) [inline scripts/tuning.py]:
```suggestion
        "max_concurrent_trials": args.num_nodes * 2,  # less than GPUs per node (4) avoid OOM
```

**Comment by rogerkuou** [comment]
@SarahAlidoost thanks! I addressed your comments. shall we merge this PR now? Or do you want to wait till we perform a training with multiple months of watervapor?

**Review by SarahAlidoost — CHANGES_REQUESTED** [review]
@rogerkuou thanks for addressing the comments. I'm reviewing the results. Meanwhile, there are several issues in the notebooks that should be fixed before merging. I couldn't comment directly on the notebooks, so I wrote them here:

1. In daily notebooks, under section "**Prepare the test data**", (in sst notebook see cell 20 and in watervapor notebook see cell 21), the variable `spatial_patch_size` should be replaced with `spatial_crop_size`. 
2. In hourly watervapor notebook, the cell containing `prediction_config = PredictionConfig(...)` is duplicated. 
3. There are unused import `add_month_day_dims` in hourly notebooks.
4. For comparison purposes, we want same data in daily and hourly notebook. Please use month 01 in hourly notebook. 
5. The variable name `lsm_mask_sst` is confusing as land-sea mask is not related to the climate variable like sst or water vapour. Please rename it. 
6. Some training configs are changed in hourly notebook, like `optimizer_lr`. Can you fix them so notebooks are consistent.

**Comment by SarahAlidoost** [comment]
> Meanwhile, following our discussion today, I will make another test on the training dataset to see if the training is able to overfit on the training data. This can validate if the problem is actually training on one month

> @rogerkuou Hi, I wonder if you have done the test on the training dataset. 

Oh, it seems that I missed the update in [your comment](https://github.com/ESMValGroup/ClimaNet/pull/74#issuecomment-5584533969).

**Comment by rogerkuou** [comment]
Hi @SarahAlidoost , thanks! I addressed your new comments:

1. In daily notebooks, under section "**Prepare the test data**", (in sst notebook see cell 20 and in watervapor notebook see cell 21), the variable `spatial_patch_size` should be replaced with `spatial_crop_size`. -> Updated the variable name
2. In hourly watervapor notebook, the cell containing `prediction_config = PredictionConfig(...)` is duplicated. -> removed duplicated cell
3. There is an unused import `add_month_day_dims` in hourly notebooks. -> removed unused import
4. For comparison purposes, we want same data in daily and hourly notebook. Please use month 01 in hourly notebook. changed to 01, need to be re-run, see item 7
5. The variable name `lsm_mask_sst` is confusing as land-sea mask is not related to the climate variable like sst or water vapour. Please rename it. -> renamed to `lsm_mask_025` and added comments
6. Some training configs are changed in hourly notebook, like `optimizer_lr`. Could you fix them so notebooks are consistent. -> changed modified parameters back: patience=10, optimizer_lr=1e-3, verbose_epoch_interval=20
7. Rerun hourly notebook on Levante -> hourly notebook has been rerun

**Comment by SarahAlidoost** [comment]
> (Updated on SEP 10) Reagrding the small trial of validate/test only on training data, see this [GH Gist](https://gist.github.com/rogerkuou/d3bc63fd444129d71a4532b4d907f0f2). It should the traning loss continue to drop, and the prediction results are significantly improved comparing to baseline. 

The dropping loss and the comparison with the baseline are good signs that the model is learning from the data. But the train loss is still quite high (0.049759 after 100 epochs). My guess is that the training subset is small compared to the model size. 

For the validation set, maybe using the full training dataset could help with the fitting issue. So for now, I’d say the results in the notebook are mainly for experimentation, and we probably shouldn’t draw any conclusions about the final error yet.

**Review by SarahAlidoost — APPROVED** [review]
@rogerkuou thanks for addressing the comments 👍 . Feel free to merge. 

One small thing for future PRs: please run Ruff only on the files changed for the PR. For example, I don’t think we need to fix `tuning.py` as part of this PR.

**Comment by rogerkuou** [comment]
> @rogerkuou thanks for addressing the comments 👍 . Feel free to merge.
> 
> One small thing for future PRs: please run Ruff only on the files changed for the PR. For example, I don’t think we need to fix `tuning.py` as part of this PR.

Thanks and noted! Do you know a practical way to do this? My formatting is done by pre-commit hooks. Maybe I will turn it off next time if I see irrelevant files. This time the formatting is introduced in the process of fixing conflicts.

Will merge this PR.
